concrete-core/src/backends/core/private/crypto/ggsw/fourier.rs

```diff
@@ -0,0 +1,725 @@
+        // the result in the output_fft_buffer variable.
+        let mut decomposition = decomposer.decompose_tensor(rounded_input_glwe);
+        // We loop through the levels (we reverse to match the order of the decomposition iterator.)
+        for ggsw_decomp_matrix in self.level_matrix_iter().rev() {
+            // We retrieve the decomposition of this level.
+            let glwe_decomp_term = decomposition.next_term().unwrap();
+            debug_assert_eq!(
+                ggsw_decomp_matrix.decomposition_level(),
+                glwe_decomp_term.level()
+            );
+            // For each levels we have to add the result of the vector-matrix product between the
+            // decomposition of the glwe, and the ggsw level matrix to the output. To do so, we
+            // iteratively add to the output, the product between every lines of the matrix, and
+            // the corresponding (scalar) polynomial in the glwe decomposition:
+            //
+            //                ggsw_mat                        ggsw_mat
+            //   glwe_dec   | - - - - | <        glwe_dec   | - - - - |
+            //  | - - - | x | - - - - |         | - - - | x | - - - - | <
+            //    ^         | - - - - |             ^       | - - - - |
+            //
+            //        t = 1                           t = 2                     ...
+            // When possible we iterate two times in a row, to benefit from the fact that fft can
+            // transform two polynomials at once.
+            let mut iterator = zip!(
+                ggsw_decomp_matrix.row_iter(),
+                glwe_decomp_term
+                    .as_tensor()
+                    .subtensor_iter(self.poly_size.0)
+                    .map(Polynomial::from_tensor)
+            );
+
+            //---------------------------------------------------------------- VECTOR-MATRIX PRODUCT
+            loop {
+                match (iterator.next(), iterator.next()) {
+                    // Two iterates are available, we use the fast fft.
+                    (Some(first), Some(second)) => {
+                        // We unpack the iterator values
+                        let zip_args!(first_ggsw_row, first_glwe_poly) = first;
+                        let zip_args!(second_ggsw_row, second_glwe_poly) = second;
+                        // We perform the forward fft transform for the glwe polynomials
+                        fft.forward_two_as_integer(
+                            first_fft_buffer,
+                            second_fft_buffer,
+                            &first_glwe_poly,
+                            &second_glwe_poly,
+                        );
+                        // Now we loop through the polynomials of the output, and add the
+                        // corresponding product of polynomials.
+                        let iterator = zip!(
+                            first_ggsw_row
```

Comment:
I do not think so. Here, polynomials are iterated over two by two because of some optimization in the FFT; does not mean that the mask has to be of length one.

Reply:
Does that answer your question @ilaria-zama?

Reply:
Yes, thanks :)